File: 301-Remove-Invalid-Parentheses/solution.java
public class Solution {
    public List<String> removeInvalidParentheses(String s) {
        List<String> result = new ArrayList<String>();
        Queue<String> queue = new LinkedList<String>();
        Set<String> visited = new HashSet<String>();
        queue.add(s);
        visited.add(s);
        boolean found = false;
        while(!queue.isEmpty()) {
            String cur = queue.poll();
            //visited.add(cur);
            if(isValid(cur)) {
                result.add(cur);
                found = true;
            }
            if(found) continue;
            for(int i = 0; i < cur.length(); i++) {
                if(cur.charAt(i) == '(' || cur.charAt(i) == ')') {
                    String temp = cur.substring(0, i) + cur.substring(i + 1);
                    if(!visited.contains(temp)) {
                        queue.add(temp);
                        visited.add(temp);
                    }
                }
            }
        }
        return result;
    }
    
    boolean isValid(String cur) {
        int left = 0, right = 0;
        for(int i = 0; i < cur.length(); i++) {
            if(cur.charAt(i) == '(') left++;
            if(cur.charAt(i) == ')') right++;
            if(left < right) return false;
        }
        return left == right;
    }
}

File: 301-Remove-Invalid-Parentheses/solution.py
class Solution(object):
    def removeInvalidParentheses(self, s):
        """
        :type s: str
        :rtype: List[str]
        """
        result, queue, visited = [], [s], set()
        visited.add(s)
        found = False
        while queue:
            cur = queue.pop(0)
            if self.isValid(cur):
                found = True
                result.append(cur)
            if found:
                continue
            for i in range(len(cur)):
                if cur[i] == '(' or cur[i] == ')':
                    temp = cur[0: i] + (cur[i + 1:] if i < len(cur) - 1 else "")
                    if temp not in visited:
                        visited.add(temp)
                        queue.append(temp)
                        found = False
        return result
    
    def isValid(self, cur):
        left, right = 0, 0
        for i in range(len(cur)):
            if cur[i] == '(':
                left += 1 
            elif cur[i] == ')':
                right += 1
            if left < right:
                return False
        return left == right
    
    
    